message/validation/consensus_validation.go

```diff
@@ -228,50 +287,65 @@ func (mv *messageValidator) validateQBFTMessageByDutyLogic(
 	return nil
 }
 
-func (mv *messageValidator) updateConsensusState(signedSSVMessage *spectypes.SignedSSVMessage, consensusMessage *specqbft.Message, state *consensusState) {
+func (mv *messageValidator) updateConsensusState(signedSSVMessage *spectypes.SignedSSVMessage, consensusMessage *specqbft.Message, consensusState *consensusState) {
 	msgSlot := phase0.Slot(consensusMessage.Height)
 
 	for _, signer := range signedSSVMessage.GetOperatorIDs() {
-		signerState := state.GetSignerState(signer)
-		if signerState == nil {
-			signerState = state.CreateSignerState(signer)
-		}
-		if msgSlot > signerState.Slot {
-			newEpoch := mv.netCfg.Beacon.EstimatedEpochAtSlot(msgSlot) > mv.netCfg.Beacon.EstimatedEpochAtSlot(signerState.Slot)
-			signerState.ResetSlot(msgSlot, consensusMessage.Round, newEpoch)
-		} else if msgSlot == signerState.Slot && consensusMessage.Round > signerState.Round {
+		var signerState *SignerState
+
+		stateBySlot := consensusState.GetOrCreate(signer)
+		signerStateInterface, ok := stateBySlot.Get(msgSlot)
+		if !ok {
+			signerState = &SignerState{}
+			signerState.Init()
 			signerState.ResetRound(consensusMessage.Round)
-		}
 
-		if len(signedSSVMessage.FullData) != 0 && signerState.ProposalData == nil {
-			signerState.ProposalData = signedSSVMessage.FullData
+			stateBySlot.Put(msgSlot, signerState)
+			mv.pruneOldSlots(stateBySlot, msgSlot)
+		} else {
+			signerState = signerStateInterface.(*SignerState)
+			if consensusMessage.Round > signerState.Round {
+				signerState.ResetRound(consensusMessage.Round)
+			}
 		}
 
-		signerCount := len(signedSSVMessage.GetOperatorIDs())
-		if signerCount > 1 {
-			signerState.SeenDecidedLengths[signerCount] = struct{}{}
+		mv.processSignerState(signedSSVMessage, consensusMessage, signerState)
+	}
+}
+
+func (mv *messageValidator) processSignerState(signedSSVMessage *spectypes.SignedSSVMessage, consensusMessage *specqbft.Message, signerState *SignerState) {
+	if len(signedSSVMessage.FullData) != 0 && consensusMessage.MsgType == specqbft.ProposalMsgType {
+		signerState.ProposalData = signedSSVMessage.FullData
+	}
+
+	signerCount := len(signedSSVMessage.GetOperatorIDs())
+	if signerCount > 1 {
+		encodedOperators, err := encodeOperators(signedSSVMessage.GetOperatorIDs())
+		if err != nil {
+			return
 		}
 
-		signerState.MessageCounts.RecordConsensusMessage(signedSSVMessage, consensusMessage)
+		signerState.SeenSigners[string(encodedOperators)] = struct{}{}
```

Comment:
this is a security concern IMO @GalRogozinski @MatheusFranco99 @nkryuchkov @moshe-blox 
if somehow a malicious signer was able to create a operatorID that is unable to be encoded, it might escape from getting into the `SeenSigners` list and send endless messages.
I might be missing an earlier check that prevents this though.

Reply:
`(*bytes.Buffer).Write` never returns an error, but I think it might be better to reject a message in this case for easier understanding. Also, theoretically, `bytes.Buffer` implementation may change in the future or we may decide to change the encoding and forget about the error handling

Reply:
done rejecting